File: python/polymorphism.py
class Polymorphism:
    def __init__(self):
        self.implementation = None

    def set_implementation(self, implementation):
        self.implementation = implementation

    def method1(self):
        print("Polymorphism: method1")

    def method2(self):
        print("Polymorphism: method2")
    
    def method2(self, a):
        print("Polymorphism: method2(a)")

class App(Polymorphism):
    def method1(self):
        print("App: method1")


class App2(Polymorphism):
    def method1(self):
        print("App2: method1")


if __name__ == "__main__":
    print("App Started")
    # app = Polymorphism()
    # choice = 0
    # while True:
    #     choice = int(input())
    #     if choice == 1:
    #         app.set_implementation(App())
    #     elif choice == 2:
    #         app.set_implementation(App2())
    #     else:
    #         break
    #     app.implementation.method1()
    print("App Ended")
    

File: OOP/c++/polymorphism.cpp
#include <iostream>

class Polymorphism {
   public:
    // virtual function
    virtual void method1() {
        std::cout << "Polymorphism: method1()" << std::endl;
    }

    virtual void method2() {
        std::cout << "Polymorphism: method2()" << std::endl;
    }

    virtual void method2(int a) {
        std::cout << "Polymorphism: method2(int a)" << std::endl;
    }
};

class App : public Polymorphism {
   public:
    void method1() {
        std::cout << "App: method1()" << std::endl;
    }
};

class App2 : public Polymorphism {
   public:
    void method1() {
        std::cout << "App2: method1()" << std::endl;
    }
};

int main() {
    std::cout << "App Started" << std::endl;
    // int choice = 0;
    // Polymorphism *app;
    // while (true) {
    //     std::cin >> choice;
    //     if (choice == 1)
    //         app = new App();
    //     else if (choice == 2)
    //         app = new App2();
    //     else {
    //         std::cout << "Invalid Choice" << std::endl;
    //         break;
    //     }
    //     app->method1();
    //     app->method2();
    //     app->method2(10);
    //     delete app;
    // }
    std::cout << "App Ended" << std::endl;
    return 0;
}